Packages: Import i18n-calypso into Calypso, take 3

blowery opened the pull request:
Take three on importing i18n-calypso into calypso's monorepo.

Previously #27123, closes #27489 (supersedes)

Very much in progress

- [x] Import existing git commits from i18n-calypso with flatten
- [ ] Migrate to ES6 / Imports
- [ ] Migrate tests to jest
- [ ] Migrate structure to src / dist to match builder
- [ ] Push new devDeps to root, if necessary
- [ ] Match up deps with calypso root
- [ ] Squash packages-prep commits
- [ ] Merge (no squash)

Tug committed:
First version, extracted from wp-calypso i18n mixin

Tug committed:
Update Readme files

Tug committed:
Renaming the project from i18n to i18n-calypso to match the npm package name

Tug committed:
Allow empty initialize() call

Tug committed:
Don't change if same locale as current, except for the default locale

Tug committed:
Bump version

Tug committed:
Remove confusing method initialize(), simplify setLocale() and assumes localeData[''].localeSlug is always given

Tug committed:
Allow empty localeData param to be given to setLocale (for default locale)

Tug committed:
Bump version

Tug committed:
Export method for simpler integration with wp-calypso

Tug committed:
Bump version

Tug committed:
Lockdown dependencies

Tug committed:
Move localize and mixin into their own npm package

Tug committed:
Fix lodash.assign dependency

Tug committed:
Bump version

Tug committed:
Add off shortcut to remove listeners

Tug committed:
Update Readme and version bump

Tug committed:
Add emit method to exposed module

Tug committed:
Include mixin and localize inside this library

Tug committed:
Update documentation link for misuse of the lib

Tug committed:
Do not overwrite moment data on load - use bundeled locale files.

- Keep building number format strings, but use strings from jed object rather than custom location.
- Stop adding momentjs strings to translation file. Use wp core's number format strings only
- For the tests: explicitly set momentjs to 'de' so it loads the bundeled data. Also add a few more tests.

props to @yoavf for this change

Tug committed:
Bump package version

Tug committed:
Include glotpress-js cli tool to extract translatable string from a js project

Tug committed:
Bump package version

Tug committed:
Add missing Readme file for the CLI

Tug committed:
Remove empty lines in generated PHP file

Tug committed:
Bump package version

Tug committed:
Print the list of input files when using the CLI

Tug committed:
Bump package version

Tug committed:
Downgrade jstimezonedetect

Tug committed:
Bump package version

Tug committed:
Remove extra line break in POT headers

Tug committed:
Bump package version

Tug committed:
Use bundeled momentjs data instead of our own

Tug committed:
Bump package version

Tug committed:
Fix tests

Tug committed:
Bump package version

Tug committed:
Fix parsing imput files when passing arguments without -i prefix

Tug committed:
Bump package version

Tug committed:
Always use double quotes in generated formats

Tug committed:
Bump package version to 1.1.0

ockham committed:
package.json: Bump React deps to ^15.1.0, still accepting 0.14.8

Tug committed:
Bump package version to 1.2.0

ockham committed:
Merge pull request #4 from Automattic/fix/tests-for-react-15

Tests: Update for React 15

ockham committed:
Merge pull request #5 from Automattic/upgrade/interpolate-components-1-1-0

Bump interpolate-components to 1.1.0

Tug committed:
Bump package version to 1.3.0

Tug committed:
Update react-test-env to allow React 15.x.y as a peer dependency

ockham committed:
Fix TypeError: `interpolateComponents is not a function` (#6)

* Fix TypeError: `interpolateComponents is not a function`

* Bump package version to 1.3.1

zinigor committed:
Merge pull request #7 from Automattic/add/textdomain-option

PHP formatter: added a textdomain option.

Tug committed:
Bump package version to 1.3.2

Tug committed:
Add a test before unmounting Localized Components so it can fail safely

Tug committed:
Bump package version to v1.4.1

akirk committed:
Improve context explanation

akirk committed:
Re-order explanations to make context less prominent

akirk committed:
Merge pull request #9 from Automattic/fix/readme-typo

Fix typo in Readme

akirk committed:
As per the spec, add verbatim line breaks to the header (#11)

Tug committed:
Bump package version to v1.4.2

akirk committed:
Include locations of the strings in the output (#12)

* Include line numbers in output

* Fix use of global variable and refactor code in pot formatter

Tug committed:
Bump package version to v1.5.0

akirk committed:
Enable POT generation from the Calypso source (#13)

* Update match parsing for the parser switch in xgettext-js.

* As per spec: just group based on singular form and context and create a new test

* Fix template literals, remove babel dependency for tests

yoavf committed:
Allow adding translations to an existing jed object (#10)

akirk committed:
Bump version to 1.6.0

akirk committed:
Add a changelog

akirk committed:
Fix POT rendering of line breaks (#14)

akirk committed:
Fixup trailing backslash replacing

Tug committed:
Merge pull request #16 from Automattic/add/glob

Add support for glob in inputFiles

Tug committed:
Bump package version to v1.6.3

Tug committed:
Merge pull request #18 from Automattic/update/lru-implementation

Replace LRU cache implementation

deBhal committed:
Merge pull request #24 from Automattic/update/broken-link

Fix cli link in README

Tug committed:
Merge pull request #25 from Automattic/add/localize-expose

Allow exposed composed component

yoavf committed:
Merge pull request #29 from edo-ran/master

Change Localize HOC display name to be consistent with react-redux

yoavf committed:
Merge pull request #30 from Automattic/update/changelog

Update change log and bump version

Tug committed:
Merge pull request #31 from Automattic/update/moment-timezone

Update moment-timezone to latest available version

Tug committed:
Merge pull request #33 from ryerh/bugfix/fix-readme-typo

Doc: fix typo and remove trailing spaces

Tug committed:
Merge pull request #32 from ryerh/master

Cli: enable keywords option

Tug committed:
Merge pull request #36 from Automattic/add/cli-globs

Added support for input file patterns in the CLI script

Tug committed:
Bump package version to v1.7.3

yoavf committed:
handle options with new plural syntax

Tug committed:
Bump package version to v1.7.4

Tug committed:
remove calypso-string.pot

nylen committed:
Improve an example

Props @tobifjellner - "such a phrase really needs to be handled with _n(). Otherwise Russian translators (among others) will have issues getting the plurals right."

Tug committed:
Merge pull request #15 from Automattic/add/php-template-literal-test

Add PHP template literal test

Tug committed:
Merge pull request #34 from ryerh/bugfix/long-line-without-seperators

Bugfix: fix call stack exceeded by long line without seperators

Tug committed:
Merge pull request #40 from Automattic/fix/mutiple-locale-sources

Merge locale data with existing one when we're not switching locale

akirk committed:
Fix improperly escaped control sequence in POT files (#41)

akirk committed:
Update changelog and version bump

ockham committed:
Update testing related dependencies

* Remove deprecated react-addons-test-utils dependency
* Upgrade enzyme, react, add react-test-renderer

samouri committed:
createClass --> es6 class (#43)

* upgrade React.createClass to use es6 classes and allow higher version of React

* bump version to 1.8.2

* modify style for class declaration

* specify node version for circleci

* update changelog

* allow v16 of react-dom

samouri committed:
Use react-create-class for es5 compat (#44)

* Use react-create-class for es5 compat

* update changelog and ver number

* move react back to reg deps

ockham committed:
Bump interpolate-components to v1.1.1 (for React 16 compat) (#45)

Also bump version, Changelog

atimmer committed:
Allow newer versions of xgettext (#46)

This way babylon can be updated in downstream projects.

akirk committed:
Fix tests for 1.8.5 (#47)

* Fix tests for 1.8.5

* Use shallow() instead of mount()

* Remove unnecessary use of setupEnzymeAdapter

* Require React 16 for tests

jsnajdr committed:
Optimize the getJedArgs function (#50)

* Optimize the getJedArgs function

Instead of creating an object with all variants and then throwing away all but one
of them, use a more efficient switch statement. Should deliver some speedup of a
function that's called thousands of times.

* Return at the end of the function instead of switch/default

akirk committed:
Add support for key hashing (#49)

* Add support for key hashing and add a hasTranslation function

* Bump version, update docs

* Require React16 for testing

* Add support for variable hash lengths

zinigor committed:
Fixed undefined iterator. (#52)

* Fixed undefined iterator.

This fixes a problem where trying to use i18n-calypso as a module in a built application would result in an `i is undefined` fatal error.

* Remove unreferenced variable

* Bump version

sirreal committed:
Update xgettext-js (#55)

* Update xgettext-js
* Add package-lock

sirreal committed:
Update to circlci v2 configuration (#57)

sirreal committed:
Update npm audit vulnerabilities (#56)

Update dependencies mentioned by `npm audit`:

* react-test-env@0.2.0
* debug@3.1.0
* mocha@5.2.0

ockham committed:
Drop obsolete react-test-env dependency (#58)

ockham committed:
Remove the mixin (#59)

ramonjd committed:
Merge pull request #54 from Automattic/fix/circular-json

Safe JSON stringification for cache key

sirreal committed:
Update CI image to node:10 and use npm cit (#63)

We're migrating to Node 10 across several other codebases primarily
because it includes a more up-to-date and reliable npm version.

Leverage `npm cit` shortcut that is essentially:
- npm ci (install packages) && npm t (test)

blowery committed:
Merge pull request #66 from Automattic/remove/async-dependency

Remove async dependency

ockham committed:
Package: Bump to v2.0.0 (#64)

blowery committed:
Update the parse options to support what calypso uses (#68)

* Update the parse options to support what calypso uses

Add exportDefaultFrom, exportNamespaceFrom, and dynamicImport.

* Alphabetize plugins

ryerh committed:
Display input files incrementally (#48)

blowery committed:
Switch to hash.js from sha1 (#67)

* Switch to hash.js from sha1

Hash.js pulling just sha1 is 2.3k gzipped, while sha1 is 8k. Switch to save some bytes and align with what calypso uses.

* Add tests for hashed lookups

blowery committed:
Prep for 2.0.1 release (#69)

yoavf committed:
Merge pull request #26 from Automattic/add/files-map-filter

CLI: add linesFilter option

renovate[bot] committed:
Add renovate.json (#70)

ramonjd committed:
Merge pull request #77 from Automattic/add/pass-locale-prop-to-wrapped-component

Localize: pass down locale as default prop to wrapped component

blowery committed:
Make the moment-timezone version less restrictive (#78)

* Make the moment-timezone version less restrictive

* match wordpress/date

ramonjd committed:
Prep for 2.0.2 release (#79)

Updating patch number and changelog.

renovate[bot] committed:
fix(deps): update dependency jed to v1.1.1 (#72)

* Update test data to work with Jed v1.1
* Add Changelog entry
* Set package version to 3.0.0

ockham committed:
Update version in package-lock.json

sirreal commented:
Rebased and adding a few changes

a8ci18n commented on packages/i18n-calypso/test/cli/examples/i18n-test-examples.jsx:
```diff
@@ -0,0 +1,103 @@
+
+	// providing context
+	content = i18n.translate( {
+		original: 'post',
```
:information_source: This string already exists with the following **contexts**:
- `null` (no context)
- `add new from admin bar`
- `noun`
- `post a comment`
- `post type singular name`
- `Verb, to post`

 Would it make sense to reuse one of the above?

a8ci18n commented on packages/i18n-calypso/test/cli/examples/i18n-test-examples.jsx:
```diff
@@ -0,0 +1,103 @@
+	} );
+
+	// passing string as initial argument
+	content = i18n.translate( 'post2', { context: 'verb2' } );
```
:information_source: This string doesn't exist elsewhere without a **context** (or with a different one). You can probably drop the context. If needed, you can use a translator comment instead to clarify meaning.

a8ci18n commented on packages/i18n-calypso/test/cli/examples/i18n-test-examples.jsx:
```diff
@@ -0,0 +1,103 @@
+		'and mixed \'single quotes\''
+	) }</div>;
+
+	content = this.translate( 'The string key text',
```
:information_source: This **context** is really long. Are you sure you don't want to use a translator comment instead?

a8ci18n commented on packages/i18n-calypso/test/index.js:
```diff
@@ -0,0 +1,366 @@
+		describe( 'translate with context', function() {
+			it( 'should find a string with context', function() {
+				assert.equal( 'translation3', translate( {
+					original: 'test3',
```
:information_source: This string doesn't exist elsewhere without a **context** (or with a different one). You can probably drop the context. If needed, you can use a translator comment instead to clarify meaning.

blowery commented on packages/i18n-calypso/package.json:
```diff
@@ -1,48 +1,45 @@
+	"module": "dist/esm/index.js",
+	"repository": {
+		"type": "git",
+		"url": "git+https://github.com/Automattic/i18n-calypso.git"
```
needs an update

blowery commented on packages/i18n-calypso/package.json:
```diff
@@ -1,48 +1,45 @@
+		"i18n-calypso": "bin/index.js"
+	},
+	"bugs": {
+		"url": "https://github.com/Automattic/i18n-calypso/issues"
```
update

blowery commented on packages/i18n-calypso/package.json:
```diff
@@ -1,48 +1,45 @@
+	"bugs": {
+		"url": "https://github.com/Automattic/i18n-calypso/issues"
+	},
+	"homepage": "https://github.com/Automattic/i18n-calypso#readme",
```
update

sirreal committed:
Prepare i18n package for Calypso repo

In order to preserve the project history, all commits have been
transferred to this branch and should be merged into Calypso master.

However, there is significant work to be done before this package is up
to Calypso standards and ready to replace the currently used package.

This commit allows us to move the code into the repository without
changing Calypso's current behavior. It will continue to use the
published package.

- Ignore i18n package from lerna
- Ignore i18n-calypso tests
  Temporarily disable tests so package can be merged maintaining history,
  then iterative changes can be made as follow-up.
- Rename i18n-calypso in package.json to disable lerna
  Prevent lerna from detecting the module locally and symlinking it for
  use from Calypso by changing the package name.
- Ignore i18n tests from translation

sirreal commented:
i18n job should be fixed by #29260 
I'd like to ignore the tests as well so this can be merged preserving the history. Initially, only the source would move into Calypso without being used directly, running tests, or any significant impact on Calypso.

The cleanup and prep can occur in #29229 and other follow-ups. There's a significant amount for this particular package.

sirreal commented:
I've disabled tests and cherry-picked the i18n fix. Any objections to landing this and doing iterative cleanup in follow-ups?

sirreal commented:
Rebased, i18n fix is now on master.

blowery commented:
Looks like this is still symlinking in i18n-calypso and breaking calypso.

blowery commented:
I think if we want to land this, we'll need to temporarily change the package name to prevent lerna from symlinking it to calypso.

sirreal commented:
> I think if we want to land this, we'll need to temporarily change the package name to prevent lerna from symlinking it to calypso.

Gave that a spin in 87801923ccd39f43848306c66121fd1ea1737c28

blowery commented:
@sirreal ` Error: Invalid package name "__i18n-calypso": name cannot start with an underscore`

sirreal commented on lerna.json:
```diff
@@ -2,8 +2,9 @@
-		"**/package-lock.json"
+		"**/.eslintrc.{js,json,yaml,yml}",
+		"**/package-lock.json",
+		"packages/i18n-calypso"
```
Fix in follow-up

sirreal commented on packages/i18n-calypso/package.json:
```diff
@@ -0,0 +1,48 @@
+{
+  "name": "not-i18n-calypso-itreallyis",
```
Fix in follow-up

blowery replied:
nice. :)

might need to run update-deps too?

a8ci18n commented on packages/i18n-calypso/_test/cli/examples/i18n-test-examples.jsx:
```diff
@@ -0,0 +1,103 @@
+
+	// providing context
+	content = i18n.translate( {
+		original: 'post',
```
:information_source: This string already exists with the following **contexts**:
- `null` (no context)
- `add new from admin bar`
- `noun`
- `post a comment`
- `post type singular name`
- `Verb, to post`

 Would it make sense to reuse one of the above?

a8ci18n commented on packages/i18n-calypso/_test/cli/examples/i18n-test-examples.jsx:
```diff
@@ -0,0 +1,103 @@
+	} );
+
+	// passing string as initial argument
+	content = i18n.translate( 'post2', { context: 'verb2' } );
```
:information_source: This string doesn't exist elsewhere without a **context** (or with a different one). You can probably drop the context. If needed, you can use a translator comment instead to clarify meaning.

a8ci18n commented on packages/i18n-calypso/_test/cli/examples/i18n-test-examples.jsx:
```diff
@@ -0,0 +1,103 @@
+		'and mixed \'single quotes\''
+	) }</div>;
+
+	content = this.translate( 'The string key text',
```
:information_source: This **context** is really long. Are you sure you don't want to use a translator comment instead?

a8ci18n commented on packages/i18n-calypso/_test/index.js:
```diff
@@ -0,0 +1,366 @@
+		describe( 'translate with context', function() {
+			it( 'should find a string with context', function() {
+				assert.equal( 'translation3', translate( {
+					original: 'test3',
```
:information_source: This string doesn't exist elsewhere without a **context** (or with a different one). You can probably drop the context. If needed, you can use a translator comment instead to clarify meaning.

blowery commented:
do we need to ignore the translations in the test directory?

sirreal commented:
> do we need to ignore the translations in the test directory?

That's probably a good idea

sirreal commented on package.json:
```diff
@@ -271,7 +271,7 @@
     "test-server:coverage": "npm run -s test-server -- --coverage",
     "test-server:watch": "npm run -s test-server -- --watch",
-    "translate": "i18n-calypso --format pot --output-file ./calypso-strings.pot -k translate,__,_x,_n,_nx -e date \"**/*.js\" \"**/*.jsx\" \"!build/**\" \"!public/**\" \"!node_modules/**\" \"!**/node_modules/**\"",
+    "translate": "i18n-calypso --format pot --output-file ./calypso-strings.pot -k translate,__,_x,_n,_nx -e date \"**/*.js\" \"**/*.jsx\" \"!build/**\" \"!public/**\" \"!node_modules/**\" \"!**/node_modules/**\" \"!packages/i18n-calypso/_test/**\" # FIXME: when tests are restored",
```
Update when restoring tests

sirreal reviewed (approved):
Tests and full e2e are all green. This shouldn't modify any Calypso behavior. Let's 🚢!

ockham commented:
:clap: Thanks for bringing this home, folks!

sirreal commented:
> 👏 Thanks for bringing this home, folks!

It's not quite home yet, need to land #29229, probably in a series of 2-3 incremental PRs
